src/encord_active/lib/metrics/acquisition_functions.py

```diff
@@ -0,0 +1,169 @@
+        model=None,
+    ):
+        self._model = model
+        super().__init__(
+            title, short_description, long_description, metric_type, data_type, annotation_type, embedding_type
+        )
+
+    def execute(self, iterator: Iterator, writer: CSVMetricWriter):
+        for _, img_pth in iterator.iterate(desc=f"Running {self.metadata.title} acquisition function"):
+            if img_pth is None:
+                continue
+            img = self.read_image(img_pth)
+            if img is None:
+                continue
+            pred_probs = self.get_predicted_class_probabilities(img)
+            if pred_probs is None:
+                continue
+            score = self.score_predicted_class_probabilities(pred_probs)
+            writer.write(score)
+
+    def get_predicted_class_probabilities(self, image) -> Optional[np.ndarray]:
+        image_array = np.asarray(image).flatten()
+        return self._model.predict_proba([image_array])
+
+    def read_image(self, image_path: Path) -> Optional[Any]:
+        return Image.open(image_path)
+
+    @abstractmethod
+    def score_predicted_class_probabilities(self, predictions: np.ndarray) -> float:
+        pass
+
+
+class Entropy(AcquisitionFunction):
+    def __init__(self, model):
+        super().__init__(
+            title="Entropy",
+            short_description="Ranks images by their entropy.",
+            long_description=(
+                "Ranks images by their entropy. \n \n"
+                "In information theory, the **entropy** of a random variable is the average level of “information”, "
+                "“surprise”, or “uncertainty” inherent to the variable's possible outcomes. "
+                "The higher the entropy, the more “uncertain” the variable outcome. \n \n"
+                r"The mathematical formula of entropy is: $H(p) = -\sum_{i=1}^{n} p_i \log_{2}{p_i}$"
+                " \n \nIt can be employed to define a heuristic that measures a model’s uncertainty about the classes "
+                "in an image using the average of the entropies of the model-predicted class probabilities in the "
+                "image. Like before, the higher the image's score, the more “confused” the model is. "
+                "As a result, data samples with higher entropy score should be offered for annotation."
+            ),
+            metric_type=MetricType.HEURISTIC,
+            data_type=DataType.IMAGE,
+            annotation_type=AnnotationType.NONE,
+            model=model,
+        )
+
+    def score_predicted_class_probabilities(self, predictions: np.ndarray) -> float:
+        # silence divide by zero warning as the result will be correct (log2(0) is -inf, when multiplied by 0 gives 0)
+        # raise exception if invalid (negative) values are found in the predictions
+        with np.errstate(divide="ignore", invalid="raise"):
+            return -np.multiply(predictions, np.nan_to_num(np.log2(predictions))).sum(axis=1).mean()
```

Comment:
It seems like this context manager only reports a generic error like `FloatingPointError`. It would be nice to report to the user explicitly that there are negative values in the probabilities.